### ISSUE TITLE
Allow output path relative to function config file

### ISSUE THREAD

**Issue by mgoltzsche** [kptdev/kpt#1531]
### Is your feature request related to a problem? Please describe.

Users expect output paths declared within a function to be relative to the function's config file.
This came up in https://github.com/mgoltzsche/khelm/issues/12

### Describe the solution you'd like

kpt should handle function output paths relative to the function config's path when writing the output to disk.

### Describe alternatives you've considered

Manage all functions within a root directory and always call `kpt fn run .` within that directory.

### Additional context

See https://github.com/mgoltzsche/khelm/issues/12

cc @howardjohn

**Comment by mikebz** [comment]
@droot @mengqiy could be an interesting use case for you to adopt and analyze as you are working on hydration/functions v1.  khelm is a solution that is delivered in several ways including a kpt function.

**Comment by droot** [comment]
@mgoltzsche we are introducing new primitive called `pipeline` (part of `Kptfile) to run `kpt-functions` declaratively where each pkg (and subpackage) can have their own `pipeline`. In the new workflow, output generated by a function will be saved in the pkg that includes the `pipeline` that invoked that function. I think that should address this requirement.
This is in active development on `next` branch and will be in shape to be tried out in 2 weeks.

**Comment by mgoltzsche** [comment]
Sounds good! Looking forward to playing with that!

**Comment by mikebz** [comment]
@droot any chance we can link this issue to the Epic or Feature ticket?  That way when it closes we can let @mgoltzsche know?

**Comment by Shell32-Natsu** [comment]
@droot Do we have an issue to track how does pipeline handle the resources generated by function or we have implemented it?

**Comment by droot** [comment]
@mgoltzsche We have introduced new concept called pipeline of functions in `Kptfile` to run functions declaratively that should address this issue I believe. You can read more details [here](https://kpt.dev/book/04-using-functions/01-declarative-function-execution). If you want to take it for a spin, here is a link to the [installation instructions](https://kpt.dev/installation/).

**Comment by mgoltzsche** [comment]
That looks awesome! I'll try to build a khelm example with it over the weekend.

**Comment by howardjohn** [comment]
This may be FUD as I only read the docs and didn't actually get a chance to try it out, but it seems like not being to use Network or volume mounts is going to be a pretty critical issue for khelm?

**Comment by mgoltzsche** [comment]
That is true! For this to work with e.g. khelm requires to be able to declare also non-yaml files input.
I had another issue in mind that could solve this but apparently it hasn't been implemented yet: #1151

**Comment by aravindtga** [comment]
The original `kpt fn run` command was renamed to `kpt fn eval` in PR #1549 and the declarative pipeline model (`kpt fn render`) was introduced in PR #1483.

With the current design, output location is explicitly controlled:
  - `kpt fn eval DIR` - DIR defines where resources are read from and written to. `--fn-config` is purely input configuration for the function, not an output scope directive.
  - `kpt fn render` - output is scoped to the package (directory with a Kptfile). If you need output isolated to a subdirectory, make it a subpackage by adding a Kptfile there.

Output relative to a specific directory can be achieved by passing that directory as the DIR argument to `fn eval`, or by creating a subpackage for `fn render`. The ambiguity that existed with the old `kpt fn run` - which discovered function configs in subdirectories and made output scope unclear, no longer applies.

This can be closed.
